proxy/src/bind.rs

```diff
@@ -196,7 +230,112 @@ where
     type BindError = ();
 
     fn bind(&self, addr: &SocketAddr) -> Result<Self::Service, Self::BindError> {
-        Ok(self.bind.bind_service(addr, self.protocol))
+        Ok(self.bind.bind_service(addr, &self.protocol))
+    }
+}
+
+
+// ===== impl ReconstructUri =====
+
+
+impl<S> ReconstructUri<S> {
+    fn new (inner: S) -> Self {
+        Self { inner }
+    }
+}
+
+impl<S, B> tower::NewService for ReconstructUri<S>
+where
+    S: tower::NewService<
+        Request=http::Request<B>,
+        Response=HttpResponse,
+    >,
+    S::Service: tower::Service<
+        Request=http::Request<B>,
+        Response=HttpResponse,
+    >,
+    ReconstructUri<S::Service>: tower::Service,
+    B: tower_h2::Body,
+{
+    type Request = <Self::Service as tower::Service>::Request;
+    type Response = <Self::Service as tower::Service>::Response;
+    type Error = <Self::Service as tower::Service>::Error;
+    type Service = ReconstructUri<S::Service>;
+    type InitError = S::InitError;
+    type Future = Map<
+        S::Future,
+        fn(S::Service) -> ReconstructUri<S::Service>
+    >;
+    fn new_service(&self) -> Self::Future {
+        self.inner.new_service().map(ReconstructUri::new)
+    }
+}
+
+impl<S, B> tower::Service for ReconstructUri<S>
+where
+    S: tower::Service<
+        Request=http::Request<B>,
+        Response=HttpResponse,
+    >,
+    B: tower_h2::Body,
+{
+    type Request = S::Request;
+    type Response = HttpResponse;
+    type Error = S::Error;
+    type Future = Either<
+        S::Future,
+        future::FutureResult<Self::Response, Self::Error>,
+    >;
+
+    fn poll_ready(&mut self) -> Poll<(), S::Error> {
+        self.inner.poll_ready()
+    }
+
+    fn call(&mut self, mut request: S::Request) -> Self::Future {
+        if let Err(_) = h1::reconstruct_uri(&mut request) {
```

Comment:
It looks like the `ReconstructUri` service is always used, even with HTTP2 requests. In that case, it probably makes sense to check for the version here before trying to construct the `Uri` from a `Host` header.